### PR TITLE
Cors Support in MVC.

### PR DESCRIPTION
**Pull request by harshgMSFT**
Mvc part of the cors changes. This one depends on https://github.com/aspnet/CORS/pull/3
cc : @Eilon @rynowak @loudej

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 1]
```diff
@@ -80,6 +85,22 @@ protected virtual ControllerModel CreateControllerModel([NotNull] TypeInfo typeI
             AddRange(controllerModel.Filters, attributes.OfType<IFilter>());
             AddRange(controllerModel.RouteConstraints, attributes.OfType<IRouteConstraintProvider>());
 
+            var enableCors = attributes.OfType<IEnableCorsMetadata>().SingleOrDefault();
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: A scenario to consider: Enable cors for the whole app and restrict the policy to certain origins that I trust. (basically as much less work as possible for end user without using attributes maybe)

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs [hunk 2]
```diff
@@ -266,6 +270,22 @@ private bool IsIDisposableMethod(MethodInfo methodInfo, TypeInfo typeInfo)
             AddRange(actionModel.ActionConstraints, attributes.OfType<IActionConstraintMetadata>());
             AddRange(actionModel.Filters, attributes.OfType<IFilter>());
 
+            var enableCors = attributes.OfType<IEnableCorsMetadata>().SingleOrDefault();
+            if (enableCors != null)
+            {
+                var corsPolicy = _corsOptions.GetPolicy(enableCors.PolicyName);
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: Should the policy be retrieved from an implementation of ICorsPolicyProvider instead (in this case attribute-based policy provider which looks at corsoptions)...a user could replace this policy provider if they want to...

example: custom policy provider which could be attribute-based but builds the policy from information in database...

**Comment by harshgMSFT** [comment]
@lodejard @Eilon  Ping.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs [hunk 3]
```diff
@@ -31,6 +34,10 @@ public void OnProvidersExecuting([NotNull] FilterProviderContext context)
                 {
                     ProvideFilter(context, item);
                 }
+                context.Results =
+                    context.Results
+                    .OrderBy(filter => filter.Descriptor, FilterDescriptorCorsComparer.Comparer)
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: Wasn't involved in the design iterations, but it seems odd that filter providers need to be CORS-aware, no?

File: src/Microsoft.AspNet.Mvc.Core/project.json [hunk 4]
```diff
@@ -16,9 +16,10 @@
     "Microsoft.AspNet.Security.DataProtection": "1.0.0-*",
     "Microsoft.Framework.CopyOnWriteDictionary.Internal": { "version": "1.0.0-*", "type": "build" },
     "Microsoft.Framework.NotNullAttribute.Internal": { "version": "1.0.0-*", "type": "build" },
+    "Microsoft.Framework.Runtime.Interfaces": "1.0.0-*",
+    "Microsoft.AspNet.Cors.Core": "1.0.0-*",
     "Microsoft.Framework.PropertyActivator.Internal": { "version": "1.0.0-*", "type": "build" },
     "Microsoft.Framework.PropertyHelper.Internal": { "version": "1.0.0-*", "type": "build" },
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/project.json]: Extra comma on the previous line.

File: src/Microsoft.AspNet.Mvc.Core/project.json [hunk 5]
```diff
@@ -16,9 +16,10 @@
     "Microsoft.AspNet.Security.DataProtection": "1.0.0-*",
     "Microsoft.Framework.CopyOnWriteDictionary.Internal": { "version": "1.0.0-*", "type": "build" },
     "Microsoft.Framework.NotNullAttribute.Internal": { "version": "1.0.0-*", "type": "build" },
+    "Microsoft.Framework.Runtime.Interfaces": "1.0.0-*",
+    "Microsoft.AspNet.Cors.Core": "1.0.0-*",
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/project.json]: And sort

File: src/Microsoft.AspNet.Mvc/project.json [hunk 6]
```diff
@@ -8,6 +8,7 @@
         "Microsoft.AspNet.Mvc.Common": { "version": "6.0.0-*", "type": "build" },
         "Microsoft.AspNet.Mvc.Razor": "6.0.0-*",
         "Microsoft.Framework.Cache.Memory": "1.0.0-*",
+        "Microsoft.AspNet.Cors": "1.0.0-*",
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc/project.json]: Sort

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs [hunk 7]
```diff
@@ -285,6 +286,48 @@ public void IsAction_StaticMethods(string methodName)
...
+            var builder = new DefaultActionModelBuilder(null, corsOptions: mockOptions.Object);
+            var typeInfo = typeof(EnableCorsController).GetTypeInfo();
+            var method = typeInfo.GetMethod("Action");
+
+            // Act
+            var actions = builder.BuildActionModels(typeInfo, method);
+
+            // Assert
+            var action = Assert.Single(actions);
+            Assert.True(action.Filters.Any(f => f is CorsAuthorizationFilter));
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: Is this the most correct assertion? Generally `Assert.True` is frowned upon because the failure messages are very poor - there's probably a more specific assertion that could be done. But either way, should this be checking that there's exactly one filter of the type we want as opposed to see if there are 1 or more?

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs [hunk 8]
```diff
@@ -678,7 +721,7 @@ public void GetActions_MixedHttpVerbsAndRoutes_MultipleEmptyAndNonEmptyVerbs()
         {
             var options = new Mock<IOptions<AuthorizationOptions>>();
             options.Setup(o => o.Options).Returns(authOptions ?? new AuthorizationOptions());
-            return new DefaultActionModelBuilder(options.Object);
+            return new DefaultActionModelBuilder(options.Object, null);
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: Named param

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CorsTests.cs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        public async Task SuccessfulCorsRequest_AllowsCredentials_IfThePolicyAllowsCredentials()
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+
+            // Adding a custom header makes it a non simple request.
+            var requestBuilder = server
+                .CreateRequest("http://localhost/Cors/EditUserComment?userComment=abcd")
+                .AddHeader(CorsConstants.Origin, "http://foo.com")
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CorsTests.cs]: Try to avoid any possibly real-looking hostname. `example.com` is always permitted (it is a specifically reserved domain name), as well as all its sub-domains.

File: test/WebSites/CorsWebSite/project.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    },
+    "dependencies": {
+        "Kestrel": "1.0.0-*",
+        "Microsoft.AspNet.Mvc": "6.0.0-*",
+        "Microsoft.AspNet.Mvc.Xml": "6.0.0-*",
+        "Microsoft.AspNet.Mvc.TestConfiguration": "1.0.0",
+        "Microsoft.AspNet.Server.IIS": "1.0.0-*",
+        "Microsoft.AspNet.Server.WebListener": "1.0.0-*",
+        "Microsoft.AspNet.StaticFiles": "1.0.0-*",
+        "Microsoft.AspNet.Cors.Core": "1.0.0-*",
```
Comment (Eilon) [inline test/WebSites/CorsWebSite/project.json]: Sort me.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/CorsTests.cs (hunk 9 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/CorsTests.cs]: sure :+1:

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs (hunk 7 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: sure.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs (hunk 3 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: Well, it does seems odd, but this is the only place which can guarantee that the cors filter is the first ```AuthorizationFilter``` to execute. The cleaner alternative would be to have a separate filter stage but its kind of an overkill. We had discussed this at length during design review and we chose to do it this way.

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: Oh well :smile:

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 11]
```diff
@@ -31,11 +34,13 @@ public class DefaultControllerModelBuilder : IControllerModelBuilder
         public DefaultControllerModelBuilder(
             IActionModelBuilder actionModelBuilder, 
             ILoggerFactory loggerFactory,
```
Comment (lodejard) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: I think you can ask for `ILogger<DefaultControllerModelBuilder>` as a service in the constructor now, if you don't want to get the `ILoggerFactory` and call Create

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                }
+
+                if (context.HttpContext.Request.IsPreflight())
+                {
+                    // If this was a preflight, there is no need to run anything else.
+                    // Also the response is always 200 so that anyone after mvc can handle the pre filght request.
+                    context.Result = new HttpStatusCodeResult(StatusCodes.Status200OK);
+                    await Task.FromResult(true);
+                }
+                else if (!result.IsValid)
```
Comment (lodejard) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: Is this PR out-of-date compared to https://github.com/aspnet/CORS/pull/3 ?

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: yes...I will update it.

File: src/Microsoft.AspNet.Mvc/project.json (hunk 6 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc/project.json]: done.

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs (hunk 8 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: fixed.

File: test/WebSites/CorsWebSite/project.json (hunk 10 above)
Comment (harshgMSFT) [inline test/WebSites/CorsWebSite/project.json]: done.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 2 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: yeah, but that change is not in untill the middleware would be in. Will update this post that.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: :+1: on the scenario. for restricting at a global level we could add a filter at a global level and that would have restricted, however that is no different how other filters behave. I would add a func test for this nevertheless.

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs (hunk 7 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: I don't see the updated unit tests... ?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: THinking a bit more, for such a scenario it would probably make more sense to do it using the CorsMiddleware instead of registering the filter. I will add the test when middleware is in.

**Comment by harshgMSFT** [comment]
branch updated.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs [hunk 13]
```diff
@@ -20,10 +22,12 @@ namespace Microsoft.AspNet.Mvc.ApplicationModels
     public class DefaultActionModelBuilder : IActionModelBuilder
     {
         private readonly AuthorizationOptions _authorizationOptions;
+        private readonly CorsOptions _corsOptions;
 
-        public DefaultActionModelBuilder(IOptions<AuthorizationOptions> options)
+        public DefaultActionModelBuilder(IOptions<AuthorizationOptions> options, IOptions<CorsOptions> corsOptions)
         {
             _authorizationOptions = options?.Options ?? new AuthorizationOptions();
+            _corsOptions = corsOptions?.Options ?? new CorsOptions();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: I'm not sure how this got here for the `_authorizationOptions`.. Why is this here? is this in case no one ever runs a setup on the `CorsOptions`?

Can we rename `options` -> `authorizationOptions` and add a comment?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 14]
```diff
@@ -31,11 +34,13 @@ public class DefaultControllerModelBuilder : IControllerModelBuilder
         public DefaultControllerModelBuilder(
             IActionModelBuilder actionModelBuilder, 
             ILoggerFactory loggerFactory, 
-            IOptions<AuthorizationOptions> options)
+            IOptions<AuthorizationOptions> options, 
+            IOptions<CorsOptions> corsOptions)
         {
             _actionModelBuilder = actionModelBuilder;
             _logger = loggerFactory.CreateLogger<DefaultControllerModelBuilder>();
             _authorizationOptions = options?.Options ?? new AuthorizationOptions();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: rename

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        private bool IsClosestToAction(ActionDescriptor actionDescriptor)
+        {
+            // If there are multiple ICorsAuthorizationFilter which are defined at the class and
+            // at the action level, the one closest to the action overrides the others. 
+            // Since filterdescriptor collection is ordered (the last filter is the one closest to the action),
+            // we apply this constraint only if there is no ICorsAuthorizationFilter after this.
+            return actionDescriptor.FilterDescriptors.Last(
+                filter => filter.Filter is ICorsAuthorizationFilter).Filter == this;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: There's also `context.Filters` which is a little simpler to use

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        /// <inheritdoc />
+        public async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // If this filter is not closest to the action, it is not applicable.
+            if (!IsClosestToAction(context.ActionDescriptor))
+            {
+                return;
+            }
+
+            if (context.HttpContext.Request.Headers.ContainsKey(CorsConstants.Origin))
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: lots of dotting to access the request and the headers. some locals would make this much more readable.

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+                var accessControlRequestMethod = 
+                        context.HttpContext.Request.Headers.Get(CorsConstants.AccessControlRequestMethod);
+                if (string.Equals(
+                        context.HttpContext.Request.Method,
+                        CorsConstants.PreflightHttpMethod,
+                        StringComparison.Ordinal) &&
+                    accessControlRequestMethod != null)
+                {
+                    // If this was a preflight, there is no need to run anything else.
+                    // Also the response is always 200 so that anyone after mvc can handle the pre filght request.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: typo `filght`

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: Also wasn't in design iterations, but would also suggest just giving these filters a high priority order and not hardcoding this behavior. What you're doing here is **not** a guarantee, the user can add more filter providers. 

We might even have cases where someone very smart is trying to work around this behavior and has to struggle with an extra layer of 'security'.

File: src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs [hunk 18]
```diff
@@ -54,8 +55,22 @@ public bool Accept([NotNull] ActionConstraintContext context)
...
+            {
+                // Update the http method if it is preflight request.
+                var accessControlRequestMethod = request.Headers.Get(CorsConstants.AccessControlRequestMethod);
+                if (string.Equals(
+                        request.Method,
+                        CorsConstants.PreflightHttpMethod,
+                        StringComparison.Ordinal) &&
+                    accessControlRequestMethod != null)
+                {
+                    method = accessControlRequestMethod;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs]: I thought this was going to handled by middleware? Is my brain going crazy?

File: src/Microsoft.AspNet.Mvc.Core/DisableCorsAuthorizationFilter.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// An <see cref="ICorsAuthorizationFilter"/> which ensures that an action does not run for a pre-flight request.
+    /// </summary>
+    public class DisableCorsAuthorizationFilter : ICorsAuthorizationFilter
+    {
+        /// <inheritdoc />
+        public Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DisableCorsAuthorizationFilter.cs]: why async?

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs (hunk 3 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: I understand the scenario of more filter providers, but I really do not see a scenario where someone where smart needs to tweak this part. We discussed having a high priority order but that was shot down as it is easier to change. 
Absolute guarantee can be provided only by a new filter stage and we did not want that.

File: src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs (hunk 18 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs]: Nope. Middleware and MVC can exist independently.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: I don't get why we're coupling these components more if it doesn't produce some sort of concrete advantage. Risk of someone messing with this and breaking it seems super low regardless of the approach we use to code it, so why choose the design with more coupling, more pieces, more testing required, etc...

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: I will repeat my self- It is easier to messup the order using Order property than to write a new custom filter provider. 
I really do not see a scenario where a user would want to mess with this. 
In general I do agree that this is more coupled ( and has the downsides that you mentioned). Lets talk to @lodejard and see if he has an opinion on this.

File: src/Microsoft.AspNet.Mvc.Core/DisableCorsAuthorizationFilter.cs (hunk 19 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/DisableCorsAuthorizationFilter.cs]: Because the enable cors is async, this one does not have to be though, but I do not want multiple interfaces.

**Comment by harshgMSFT** [comment]
@lodejard ping.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs [hunk 20]
```diff
@@ -266,6 +268,18 @@ private bool IsIDisposableMethod(MethodInfo methodInfo, TypeInfo typeInfo)
             AddRange(actionModel.ActionConstraints, attributes.OfType<IActionConstraintMetadata>());
             AddRange(actionModel.Filters, attributes.OfType<IFilter>());
 
+            var enableCors = attributes.OfType<IEnableCorsMetadata>().SingleOrDefault();
```
Comment (lodejard) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: I noticed the authorize code only has AuthorizeAttribute, but no associated interface. Those should probably be consistent, maybe file a bug to add `IAuthorizeAttribute` along with the `AuthorizeAttribute`

Also, if we do marker interfaces in favor of a well-known attribute clas it should be named something like `IEnableCorsAttribute`.

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: https://github.com/aspnet/Security/issues/193

I will update the name to `IEnableCorsAttribute`

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+            if (!IsClosestToAction(context.Filters))
+            {
+                return;
+            }
+
+            var httpContext = context.HttpContext;
+            var request = httpContext.Request;
+            if (request.Headers.ContainsKey(CorsConstants.Origin))
+            {
+                var corsPolicyProvider = httpContext.RequestServices.GetRequiredService<ICorsPolicyProvider>();
```
Comment (lodejard) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: Move the ICorsPolicyProvider and ICorsService to constructor parameters - they don't need to vary per-request

File: src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs (hunk 18 above)
Comment (lodejard) [inline src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs]: @rynowak let's schedule a followup to review HttpMethodConstraint

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs (hunk 21 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: This is not in the ctor parameters as `CorsAuthorizationFilter` gets injected at application model build time, is your suggestion to inject these services in `DefaultControllerModelBuilder`  as well? 
BTW this is similar to how we do injection for `AuthorizationService` as well.

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: Another thing that we can potentially do is instead of creating a `CorsAuthorizationFilter ` we create a `CorsAuthorizationFilterFactory : IFilterFactory` which in its `CreateInstance` can create the  `CorsAuthorizationFilter`. 
The only downside is since we require to pass in the policy name to the constructor the factory cannot be DI ed.

File: src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 20 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: Looks like it's still waiting for a rename.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: Should this test be here now?

File: src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs [hunk 22]
```diff
@@ -1,6 +1,5 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs]: but WHY?

File: src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs (hunk 21 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs]: There's a few ways to solve this. You can get the `CorsAuthorizationFilter` from DI (transient) and then set the policy. The Factory can cache the 'real' filter

File: src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/HttpMethodConstraint.cs]: Had a discussion yesterday with @loudej - we're happy which how this looks.

File: test/WebSites/CorsWebSite/Startup.cs [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+        public void Configure(IApplicationBuilder app)
+        {
+            var configuration = app.GetTestConfiguration();
+
+            app.UseServices(services =>
+            {
+                services.AddMvc();
+                services.Configure<MvcOptions>(options =>
+                {
+                    options.AddXmlDataContractSerializerFormatter();
```
Comment (rynowak) [inline test/WebSites/CorsWebSite/Startup.cs]: why?

File: test/WebSites/CorsWebSite/Startup.cs [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+                            builder.AllowCredentials()
+                                   .AllowAnyOrigin()
+                                   .AllowAnyHeader()
+                                   .AddMethods("PUT", "POST")
+                                   .AddExposedHeaders("exposed1", "exposed2");
+                        });
+                });
+            });
+
+            app.UseErrorReporter();
```
Comment (rynowak) [inline test/WebSites/CorsWebSite/Startup.cs]: PLEASE NO. This makes it super hard to diagnose errors.

File: test/WebSites/CorsWebSite/Startup.cs [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+            app.UseMvc(routes =>
+            {
+                routes.MapRoute(
+                    name: "default",
+                    template: "{controller}/{action}/{id?}");
```
Comment (rynowak) [inline test/WebSites/CorsWebSite/Startup.cs]: Remove if unused, I think your controllers use attribute routing no?

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: Yes, I will add that.

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 20 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: yup.

**Comment by harshgMSFT** [comment]
@rynowak  updated.

File: src/Microsoft.AspNet.Mvc.Core/Filters/CorsAuthorizationFilterFactory.cs [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            get
+            {
+                return DefaultOrder.DefaultCorsSortOrder;
+            }
+        }
+
+        public IFilter CreateInstance([NotNull] IServiceProvider serviceProvider)
+        {
+            var filter = serviceProvider.GetRequiredService<ICorsAuthorizationFilter>() as CorsAuthorizationFilter;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/CorsAuthorizationFilterFactory.cs]: Never use `as` if you're not able to handle the failure case. We should just make the setter part of the interface.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs [hunk 27]
```diff
@@ -2,9 +2,12 @@
...
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
+using Microsoft.AspNet.Cors.Core;
 using Microsoft.AspNet.Mvc.Core;
 using Microsoft.Framework.Internal;
+using System.Linq;
 
 namespace Microsoft.AspNet.Mvc.Filters
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: Should be able to revert this

**Comment by rynowak** [comment]
:shipit: after fixing minor issues.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs (hunk 27 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterProvider.cs]: yeah my vs is still wonky .. cleaned up.

**Comment by harshgMSFT** [comment]
015edefa96c2eec7a474489a93a533e10ac74c7e
